types/wlr_surface_layers.c

```diff
@@ -0,0 +1,287 @@
+		wl_resource_get_user_data(resource);
+	struct wlr_surface *surface = wl_resource_get_user_data(surface_resource);
+	struct wlr_output *output = wl_resource_get_user_data(output_resource);
+
+	if (layer > SURFACE_LAYERS_LAYER_OVERLAY) {
+		wl_resource_post_error(resource, SURFACE_LAYERS_ERROR_INVALID_LAYER,
+			"Unknown layer");
+		return;
+	}
+	if (wlr_surface_set_role(surface, surface_layers_role, resource,
+			SURFACE_LAYERS_ERROR_ROLE)) {
+		return;
+	}
+
+	struct wlr_layer_surface *layer_surface =
+		calloc(1, sizeof(struct wlr_layer_surface));
+	if (layer_surface == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	layer_surface->resource = wl_resource_create(client,
+		&layer_surface_interface, wl_resource_get_version(resource), id);
+	wlr_log(L_DEBUG, "new layer_surface %p (res %p)", layer_surface,
+		layer_surface->resource);
+	wl_resource_set_implementation(layer_surface->resource, &layer_surface_impl,
+		layer_surface, layer_surface_resource_destroy);
+
+	layer_surface->surface_layers = surface_layers;
+	layer_surface->surface = surface;
+	layer_surface->output = output;
+	layer_surface->layer = layer;
+
+	layer_surface->current = calloc(1, sizeof(struct wlr_layer_surface_state));
+	if (layer_surface->current == NULL) {
+		free(layer_surface);
+		wl_client_post_no_memory(client);
+		return;
+	}
+	layer_surface->pending = calloc(1, sizeof(struct wlr_layer_surface_state));
+	if (layer_surface->pending == NULL) {
+		free(layer_surface->current);
+		free(layer_surface);
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	wl_signal_init(&layer_surface->events.destroy);
+	wl_signal_init(&layer_surface->events.commit);
+
+	wl_signal_add(&layer_surface->surface->events.destroy,
+		&layer_surface->surface_destroy_listener);
+	layer_surface->surface_destroy_listener.notify =
+		handle_layer_surface_destroy;
+	wl_signal_add(&layer_surface->surface->events.commit,
+		&layer_surface->surface_commit_listener);
+	layer_surface->surface_commit_listener.notify =
+		handle_layer_surface_commit;
+
+	bool inserted = false;
+	struct wlr_layer_surface *ls;
+	wl_list_for_each_reverse(ls, &surface_layers->surfaces, link) {
+		if (ls->layer <= layer_surface->layer) {
+			inserted = true;
+			wl_list_insert(&ls->link, &layer_surface->link);
+			break;
+		}
+	}
+	if (!inserted) {
+		wl_list_insert(&surface_layers->surfaces, &layer_surface->link);
+	}
+
+	wl_signal_emit(&surface_layers->events.new_surface, layer_surface);
+}
+
+static const struct surface_layers_interface surface_layers_impl = {
+	.get_layer_surface = surface_layers_get_layer_surface,
+};
+
+static void surface_layers_bind(struct wl_client *wl_client,
+		void *_surface_layers, uint32_t version, uint32_t id) {
+	struct wlr_screenshooter *surface_layers = _surface_layers;
+	assert(wl_client && surface_layers);
+	if (version > 1) {
+		wlr_log(L_ERROR, "Client requested unsupported surface_layers version,"
+			"disconnecting");
+		wl_client_destroy(wl_client);
```

Comment:
This will never be reached. The `wl_registry` will give a protocol error and disconnect the client for you.